aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -329,6 +311,54 @@ Tensor& index_out(Tensor& result, const Tensor & self, const torch::List<c10::op
   return result;
 }
 
+Tensor & put_(Tensor & self, const Tensor& index, const Tensor & source, const bool accumulate) {
+  // See note [Writing Nondeterministic Operations]
+  // Nondeterministic when index contains duplicate entries and we do not accumulate
+  if (!accumulate) {
+    at::globalContext().alertNotDeterministic("put_");
+  }
+
+  // Type and device checks
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "put_(): Expected a long tensor for index, but got ", index.scalar_type())
+  TORCH_CHECK(self.scalar_type() == source.scalar_type(), "put_(): self and source expected to have the same dtype, but got (self) ", self.scalar_type(), " and (source) ", source.scalar_type());
+  TORCH_CHECK(self.device() == source.device() && self.device() == index.device(),
+      "put_(): self, index and source expected to be in the same device, but got (self) ",
+      self.device(), ", (index) ", index.device(), ", and (source) ", source.device());
+
+  // index checks
+  TORCH_CHECK_INDEX(source.numel() == index.numel(), "put_(): Expected 'source' and 'index' to have the same number of elements, but got 'source' (", source.numel(), "), 'index'(", index.numel(), ")");
```

Comment:
This is divergent from numpy behavior "If v is shorter than ind it will be repeated as necessary.", but consistent with pre-porting behavior.  I don't think we should reproduce numpy behavior though, cc @mruberry

Reply:
I'm marking this as resolved. We can discuss this in the issue Mike opened https://github.com/pytorch/pytorch/issues/54116